pkg/http/server.go

```diff
@@ -123,9 +125,14 @@ func (srv *Server) Shutdown() error {
 		return errors.New("http server already in shutdown")
 	}
 
+	if atomic.LoadInt32(&srv.upAndRunning) <= 0 {
+		return nil
+	}
+
 	// Close underneath HTTP listener.
 	srv.listenerMutex.Lock()
 	err := srv.listener.Close()
```

Comment:
Checking Listener for nil was my first approach, but then I discussed it with @harshavardhana and he suggested to implement the fix this way by introducing additional field in the structure. 
As a continuation, the goal is to refactor the code more. So, the plan is to get rid of the mutex logic from Listener code/design. We can discuss it further on the phone if you'd like.